src/psi.rs

```diff
@@ -0,0 +1,276 @@
+    // TODO: those looooooong types are ugly
+    // TODO: these should be global variables
+    let key: aes::cipher::generic_array::GenericArray<u8, UInt<UInt<UInt<UInt<UInt<UTerm, B1>, B0>, B0>, B0>, B0>> = Aes128GcmSiv::generate_key(&mut OsRng);
+    let cipher: aes_gcm_siv::AesGcmSiv<aes::Aes128> = Aes128GcmSiv::new(&key);
+    let nonce: &aes::cipher::generic_array::GenericArray<u8, aes::cipher::typenum::UInt<aes::cipher::typenum::UInt<aes::cipher::typenum::UInt<aes::cipher::typenum::UInt<aes::cipher::typenum::UTerm, aes::cipher::typenum::B1>, aes::cipher::typenum::B1>, aes::cipher::typenum::B0>, aes::cipher::typenum::B0>> = Nonce::from_slice(b"unique nonce"); // 96-bits; unique per message
+
+    let input_string: String = std::format!("{input}");
+    // In AES, encryption and decryption are done by the same operation
+    // from: https://docs.rs/aes-gcm-siv/latest/aes_gcm_siv/#usage
+    // also: https://stackoverflow.com/questions/23850486/how-do-i-convert-a-string-into-a-vector-of-bytes-in-rust
+    // TODO: why does it give an error when I run `.decrypt`?
+    // TODO: is is_ok() okay (lol) or should I propagate the error with Result<T, E>?
+    // TODO: test whether applying cipher.encrypt gives you the initial value (how is this possible?)
+    let permuted_bytes = cipher.encrypt(nonce, input_string.as_bytes().as_ref());
+    assert!(permuted_bytes.is_ok());
+    let permuted: Fq = <Fq as PrimeField>::from_le_bytes_mod_order(&permuted_bytes.unwrap());
+
+    permuted
+}
+
+/// Hash function from arbitrary string to field element
+/// H_1 in the paper.
+///
+/// # Arguments
+///
+/// * `input` - An element of one of the private sets. In the future, the library may accept different element types, but currently only u64 is accepted.
+/// 
+/// # Outputs
+///
+/// * `hash` - A field element
+fn hash_1(input: u64) -> Fq {
+    // https://docs.rs/sha2/latest/sha2/
+    let mut hasher = Sha256::new();
+    hasher.update(input.to_le_bytes());
+    let result = hasher.finalize();
+
+    // need to shorten it to u64 for now.
+    // represent those 8 bytes as a single u64
+    let hash: Fq = <Fq as PrimeField>::from_le_bytes_mod_order(result[..].try_into().unwrap());
+
+    hash
+}
+
+/// Hash function from arbitrary string x field element to field element
+/// H_2 in the paper.
+///
+/// # Arguments
+///
+/// * `input1` - An element of one of the private sets. In the future, the library may accept different element types, but currently only u64 is accepted.
+/// * `input2` - A field element
+/// 
+/// # Outputs
+///
+/// * `hash` - A field element
+fn hash_2(input1: u64, input2: Fq) -> Fq {
+    // finally, hash x_i with the output from the previous step
+    let mut hasher = Sha256::new();
+    // TODO: when you stack up the `update`s, it doesn't overwrite everything except the last one, right?
+    hasher.update(input1.to_le_bytes());
+    let k_i_string: String = std::format!("{input2}");
```

Comment:
re-pinging on this unnecessary string conversion. can implement the hash trait on Fq and do it directly instead of converting to string